Framework/PythonInterface/plugins/algorithms/OrderWorkspaceHistory.py

```diff
@@ -37,70 +39,45 @@ def validateInputs(self):
-        self.log().debug("Found datetime stamp: {0}".format(date_time))
-
-        # Remove whitespace and any trailing zeros
-        stipped_time = date_time.strip().rstrip('0')
-        return datetime.datetime.strptime(stipped_time, "%Y-%m-%dT%H:%M:%S.%f")
-
-    def write_ordered_workspace_history(self):
         self.log().debug("Started saving workspace histories")
-        source_folder = self.getPropertyValue(_recovery_folder)
 
-        # Get list of all workspace histories
-        onlyfiles = [f for f in os.listdir(source_folder)
-                     if os.path.isfile(os.path.join(source_folder, f))]
-        historyfiles = [x for x in onlyfiles if x.endswith('.py')]
-
-        self.log().debug("Found {0} history files".format(len(historyfiles)))
-
-        # Read each history in as a list of tuples
-        original_lines = set()
-        for infile in historyfiles:
-            with open(os.path.join(source_folder, infile), 'r') as f:
-                for line in f:
-                    original_lines.add(line)
-
-        commands = []
-        # Remove any comment lines
-        for line in original_lines:
-            stripped_line = line.lstrip()
-            if stripped_line[0] is not '#':
-                commands.append(stripped_line)
-
-        self.log().debug("Found {0} unique commands".format(len(commands)))
-
-        # Add lists of histories together
-        all_unique_commands = [command.strip().split('#') for command in commands]
-
-        # Convert the datetime into a sortable integer
-        all_unique_commands = [(i[0], self._concatenate_iso_datetime(i[1]))
-                               for i in all_unique_commands]
-
-        # Sort the new list on datetime integer
-        all_unique_commands.sort(key=lambda time: (time[1]))
+        all_lines = []
+
+        #for fn in historyfiles:
+        for fn in glob.iglob(os.path.join(self.getPropertyValue(_recovery_folder), '*.py')):
+            with open(fn) as f:
+                tokens = tokenize.generate_tokens(f.readline)
+
+                line = None
+                for t in tokens:
+                    #Start a new line when we see a name
+                    if line is None and t[0] == tokenize.NAME:
+                        line = [t]
+                    # End the line when we see a logical line ending
+                    elif t[0] == tokenize.NEWLINE:
+                        # Only care about the line if it has a comment
+                        have_comment = any(x[0] == tokenize.COMMENT for x in line)
+                        if have_comment:
+                            all_lines.append(line)
+                        line = []
+                    # Everything in between we care about
+                    elif line is not None:
+                        line.append(t)
+
+        # l[-1][1][1:] is the comment string, with the preceeding hash stripped off
+        all_commands = [(''.join(t[1] for t in l[:-1]), l[-1][1][1:]) for l in all_lines]
```

Comment:
I think the first element of the tuple needs explaining too, without being familiar with `tokenize` you probably wouldn't realise that this is reconstructing the line minus the comment (last token).